calyx/src/analysis/share_set.rs

```diff
@@ -0,0 +1,71 @@
+use crate::ir;
+use crate::ir::RRC;
+use std::collections::HashSet;
+
+/// Stores a Hashset that contains the type names of all components and primitives 
+/// marked with either "share" or "state_share",depending on what the user wants. 
+/// Methods implemented by this struct can 
+/// be used to determine whether a given cell is shareable or not
+/// Used by `live_range_analysis.rs`, `cell_share.rs`, and `infer_share.rs`
+#[derive(Default, Clone)]
+pub struct ShareSet {
+    shareable: HashSet<ir::Id>,
+    is_state_share: bool,
+}
+
+impl ShareSet {
+    pub fn new(shareable: HashSet<ir::Id>, is_state_share: bool) -> Self {
+        ShareSet { shareable, is_state_share  }
+    }
+
+    ///Constructs a shareset from the context. Looks for "state_share" types if 
+    ///is_state_share is true, and "share" types otherwise. 
+    pub fn from_context(ctx: &ir::Context, is_state_share: bool) -> Self {
```

Comment:
One trick to use here is you expect `is_state_share` to always be a constant (either `true` or `false`) is to make the function use [`const` generics](https://blog.rust-lang.org/2021/02/26/const-generics-mvp-beta.html):
```
pub fn from_context<const IS_STATE_SHARE: bool>(ctx: &ir::Context) { ... }
```

You can use `IS_STATE_SHARE` normal in the function body and at call sites, you can do:
```
from_context::<true>(ctx)
from_context::<false>(ctx)
```

The benefit of this is the compiler will generate a specialized version for each value of `IS_STATE_SHARE` for you